facebookresearch/ParlAI
# Multiparty Chat

**Patch description**
This patch adds basic multiparty chat functionality to ParlAI.
This is achieved by relaxing the "two-agents-only" assumption in `TorchAgent` and "round-robin" speaker order determination in the `parley` method for `MultiAgentDialogWorld` to allow for multiple agents that can speak in any order in the same chat.
TorchAgent and History also now have the ability to dynamically add new speakers on the fly based on the corresponding `speaker_field` in any new observations.
To ensure backward-compatability, the new behavior only manifests when corresponding command-line flags are set.

**Testing steps**
To test for default behavior without the new functionalities, run existing unit tests/ scripts as-is.
To test for new behavior, specify these new command-line arguments:

1) For `MultiAgentDialogWorld` in `worlds.py`:
`--parley-mode=async` (enables async behavior where speakers can talk in any order)

2) For `TorchAgent`:
`--allow-multiple-observe` (relaxes the check in `self.__observe_invariants` that disallow multiple `observe` before `act`)
`--speaker-field=speaker` (set it to the field name to get the speaker id from observations)
